app/Jobs/FetchAd.php

```diff
@@ -9,108 +9,114 @@
+        if (isset($this->source)) {
+            $donePages = $this->fetch();
+        } else {
+            Log::error('source is not set with name: ' . $this->sourceName);
+            throw new Exception('source is not set with name: ' . $this->sourceName);
         }
     }
 
     /**
-     * @param stdClass $source
-     * @return array
+     * @return int
      */
-    private function fetch(stdClass $source): array
+    private function fetch(): int
     {
-        $fetchUrl = (new SourceFetchUrlGenerator($source))->generateUrl();
+        $fetchUrl = (new SourceFetchUrlGenerator($this->source, $this->since))->generateUrl();
 
-        if(is_null($fetchUrl)) {
-            return [0, 0];
+        if (is_null($fetchUrl)) {
+            return 0;
         }
 
-        $failedPages = 0;
         $donePages = 0;
 
         do {
             $counter = 0;
-            [$fetchDone , $items , $currentPage , $nextPageUrl , $lastPage, $resultText] = $this->adFetcher->fetchAd($fetchUrl);
-            if(!$fetchDone){
-                $failedPages++;
-                continue;
+            [$fetchDone, $items, $currentPage, $nextPageUrl, $lastPage, $resultText] = $this->adFetcher->fetchAd($fetchUrl);
+            if (!$fetchDone) {
+                Log::error('response status code is not 200 | request url: ' . $fetchUrl);
+                throw new Exception('response status code is not 200 | request url: ' . $fetchUrl);
             }
 
             if (empty($items)) {
-                continue;
+                Log::error('response data is null | request url: ' . $fetchUrl);
+                throw new Exception('response data is null | request url: ' . $fetchUrl);
+            }
+            $this->storeItems($this->source, $items, $counter);
+
+            if ($currentPage === 1) {
+                $this->insertFetch($this->source->id, $currentPage, $lastPage, $nextPageUrl);
+                $this->lastFetch = DB::table('fetches')->where('source_id', $this->source->id)
+                    ->latest()->first();
+            } else {
+                $this->updateFetch($this->source->id, $currentPage, $nextPageUrl, $this->lastFetch->id);
             }
 
-            $this->storeItems($source, $items, $counter);
-
-            $this->insertOrUpdateFetch($source, $currentPage, $lastPage, $nextPageUrl);
             $fetchUrl = $nextPageUrl;
             $donePages++;
-        } while ($currentPage < $lastPage );
-
-        return [$donePages, $failedPages];
-    }
-
-    /**
-     * @param stdClass $source
-     * @param $currentPage
-     * @param $lastPage
-     * @param $nextPageUrl
-     * @return bool
-     */
-    private function insertOrUpdateFetch(stdClass $source, $currentPage, $lastPage, $nextPageUrl): bool
-    {
-        if ($currentPage == 1) {
-            return $this->insertFetch($source->id, $currentPage, $lastPage, $nextPageUrl);
-        }
+            if ($currentPage === $this->lastFetch->last_page) {
```

Comment:
check in this if that failed urls are empty and if true set completed_at for fetch